### PR TITLE
ENT-11765: Changed cf-execd's sleep behavior so it attempts to wake up at the beginning of every minute (3.21)

### PR DESCRIPTION
**Pull request by olehermanse**
This could fix issues if things are slow and agent runs end up getting skipped if cf-execd wakes up too late.

(cherry picked from commit 547f3679737b695e01d7aa26c4bc48b8c3d22c01)

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom build this in Jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/10714/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/10714/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-10714/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/10715/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/10715/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-10715/

File: cf-execd/cf-execd.c [hunk 1]
```diff
@@ -424,6 +424,17 @@ void ThisAgentInit(void)
...
 }
 
+// Return 2 to 62 seconds depending on how much time is left until the next
+// minute starts. Since cf-execd wakes up "every minute" to evaluate its
+// schedule, we want to do this at the start of the minute, to avoid
+// accidentally skipping any runs if things are slow.
+static inline time_t GetPulseTime()
+{
+    const time_t current_second = (time(NULL) % 60);
+    const time_t remaining_seconds = 60 - current_second;
```
Comment (larsewi) [inline cf-execd/cf-execd.c:434]:
```suggestion
    const time_t remaining_seconds = CFPULSETIME - current_second;
```

File: cf-execd/cf-execd.c [hunk 2]
```diff
@@ -424,6 +424,17 @@ void ThisAgentInit(void)
...
+    return remaining_seconds + 2;
```
Comment (larsewi) [inline cf-execd/cf-execd.c:435]: Why `+2`? Please add comment :)

**Review by larsewi — APPROVED** [review]

File: cf-execd/cf-execd.c (hunk 1 above)
Comment (craigcomstock) [reply cf-execd/cf-execd.c:434]: but this is the cherry pick, so change should go in https://github.com/cfengine/core/pull/5498

Comment (craigcomstock) [reply cf-execd/cf-execd.c:434]: I suggested in the main PR to just remove `CFPULSETIME` https://github.com/cfengine/core/pull/5498#discussion_r1592703126

File: cf-execd/cf-execd.c (hunk 2 above)
Comment (craigcomstock) [reply cf-execd/cf-execd.c:435]: again, change these in the master PR and then re-cherry-pick right? :)

**Comment by olehermanse** [comment]
@cf-bottom jenkins again, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/10732/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/10732/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-10732/

**Pull request merged**